text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+              status_message: string | string_dictionary
+```
+
+## Implementation Recommendations
+
+### Protocol Extension and Fallback Mechanism
+
+The support of this new protocol can only be progressive, so implementers are advised to follow the following
+implementation recommendations in phase 1:
+
+* OTLP Receiver: Listen on a single TCP port for both OTLP and OTLP Arrow. The goal is to make the support of this
+  protocol extension
+  transparent and automatic. This can be achieved by adding the `ArrowStreamService` to the same gRPC listener. A
+  configuration
+  parameter will be added to the OTLP receiver to disable this default behavior to support specific uses.
+* OTLP Exporter: By default the OTLP exporter should initiate a connection to the `ArrowStreamService` endpoint of the target
+  receiver. If this connection fails because the `ArrowStreamService` is not implemented by the target, the exporter
+  must automatically fall back on the behavior of the classic OTLP protocol. A configuration parameter could be added to
+  disable this default behavior.
+
+The implementation of these two rules should allow a seamless and
+adaptive integration of OTLP Arrow into the current ecosystem
+generally.
+
+For the prototype specifically, which is a fork of the OpenTelemetry
+collector codebase, we have derived the OTLP/gRPC-Arrow exporter and
+receiver as set of changes directly to the `receiver/otlpreceiver` and
+`exporter/otlpexporter` components, with new `internal/arrow` packages
+in both.  With every collector release we merge the OTLP-Arrow changes
+with the mainline components to maintain this promise of
+compatibility.
+
+OTLP-Arrow supports conveying the gRPC metadta (i.e., http2 headers) using a dedicated `bytes` field.  Metadata is
+encoded using [hpack](https://datatracker.ietf.org/doc/rfc7541/) like a typical unary gRPC request.
+
+Specifically:
+
+#### OTLP/gRPC Receiver
+
+When Arrow is enabled, the OTLP receiver listens for both the standard unary gRPC service OTLP and OTLP-Arrow stream
+services.  Each stream uses an instance of the OTel-Arrow-Adapter's
+[Consumer](https://pkg.go.dev/github.com/f5/otel-arrow-adapter@v0.0.0-20230112224802-dafb6df21c97/pkg/otel/arrow_record#Consumer). Sets
+`client.Metadata` in the Context.
+
+#### OTLP/gRPC Exporter
+
+When Arrow is enabled, the OTLP exporter starts a fixed number of streams and repeatedly sends one `plog.Logs`,
+`ptrace.Traces`, or `pmetric.Metrics` item per stream request.  The `exporterhelper` callback first tries to get an
+available stream, blocking when none are available (or until the connection is downgraded), and then falls back to the
+standard unary gRPC path.  The stream-sending mechanism internally handles retries when failures are caused by streams
+restarting, while honoring the caller's context deadline, to avoid delays introduced by allowing these retries to go
+through the `exporterhelper` mechanism.
```

Comment:
>to avoid delays introduced by allowing these retries to go through the `exporterhelper` mechanism.

This part is unclear to me.

Reply:
The situation is one where the request has been issued on a stream and we are waiting for that stream to respond when it tells us to GoAway because it's closing down. This is seen as "no fault" associated with the data itself and all the requests that were pending on that stream are immediately retried. Again this is the [specific condition where a request is pending and the stream closes for any reason after the request was successfully sent](https://github.com/open-telemetry/experimental-arrow-collector/blob/30e0ffb230d3d2f1ad9645ec54a90bbb7b9878c2/exporter/otlpexporter/internal/arrow/stream.go#L175).

 I could pass a retryable error back to the `exporterhelper` and it would retry after a delay, but the delay would impact users because of an underlying mechanism recycling streams.